test/header/CMakeLists.txt

```diff
@@ -28,6 +27,53 @@ option (SEQAN3_FULL_HEADER_TEST "Test seqan3 headers as well as the headers of e
 #
 # \sa Modified version from Bio-Formats
 # https://github.com/openmicroscopy/bioformats/blob/d3bb33eeda23e81f78fd25f658bfc14a4363805f/cpp/cmake/HeaderTest.cmake#L81-L113
+#
+# \sa https://en.cppreference.com/w/cpp/language/storage_duration#external_linkage
+#
+# ======== Detailed Explanation ========
+#
+# `test/header/generate_header_source.cmake` generates for each given header file two `.cpp` files.
+#
+# Let header be `include/seqan3/search/fm_index/bi_fm_index_cursor.hpp` and component be `seqan3`.
+#
+# * `<build_dir>/seqan3_header_test/seqan3/search/fm_index/bi_fm_index_cursor.hpp-header-guard.cpp`
+#   which checks if the header implements a header guard.
+#
+#   This will be checked by including the header twice, if the header guard is missing the second include will produce
+#   redeclaration errors, i.e.
+#
+#   ```c++
+#   #include <seqan3/search/fm_index/bi_fm_index_cursor.hpp>
+#   #include <seqan3/search/fm_index/bi_fm_index_cursor.hpp> // redeclaration errors if bi_fm_index_cursor.hpp has no header-guards
+#   ```
+#
+# * `<build_dir>/seqan3_header_test/seqan3/search/fm_index/bi_fm_index_cursor.hpp-no-self-include.cpp`
+#   which checks if the header will not be included itself (cyclic include). For this we copy the content of the header
+#   file into the source file and if any included header includes the header itself again, we would get redeclaration
+#   errors, i.e.
+#   ```c++
+#   // ....
+#   #pragma once
+#
+#   #include <array>
+#
+#   #include <sdsl/suffix_trees.hpp>
+#
+#   #include <seqan3/alphabet/adaptation/char.hpp>
+#   #include <seqan3/alphabet/adaptation/uint.hpp>
+#   #include <seqan3/alphabet/concept.hpp>
+#   #include <seqan3/core/type_traits/range.hpp>
+#   #include <seqan3/range/views/slice.hpp>
+#   #include <seqan3/search/fm_index/fm_index.hpp> // redeclaration errors if this includes bi_fm_index_cursor.hpp
+#   #include <seqan3/search/fm_index/fm_index_cursor.hpp>
+#   #include <seqan3/std/ranges>
+#
+#   namespace seqan3
+#   // ...
+#   ```
```

Comment:
This snippet could be much shorter to make your point clear. In fact, it is rather misleading if there are so many includes. And the `#pragma once` directive must be omitted in order to see the redeclaration errors?!

I think this would be enough:
```cpp
// This is file bi_fm_index_cursor.hpp

// ...
#include <seqan3/search/fm_index/fm_index.hpp> // redeclaration errors if fm_index.hpp includes bi_fm_index_cursor.hpp
```

Reply:
Changed :)